text-underline-offset should support percentages

zakariaridouh opened the pull request:
#### 834ed1dcd01004b6fa48e5d348acba481253e88b
<pre>
text-underline-offset should support percentages
<a href="https://bugs.webkit.org/show_bug.cgi?id=263431">https://bugs.webkit.org/show_bug.cgi?id=263431</a>
<a href="https://rdar.apple.com/117246233">rdar://117246233</a>

Reviewed by Tim Nguyen.

This commit adds support for percentages in text-underline-offset [1].

Interpolation between different supported length types, such as em, %, and px
is supported and tested. Validate proper rendering with new tests. As specified
in the spec, % scales as a percentage of em/size of the font it inherits.

[1] <a href="https://drafts.csswg.org/css-text-decor-4/#valdef-text-underline-offset-percentage">https://drafts.csswg.org/css-text-decor-4/#valdef-text-underline-offset-percentage</a>

    * LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/animations/text-underline-offset-interpolation-expected.txt:
    * LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/animations/text-underline-offset-interpolation.html:
    * LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-002-expected.html:
    * LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-002.html:
    * LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-calc-expected.html: Added.
    * LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-calc-ref.html: Added.
    * LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-calc.html: Added.
    * LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-computed-expected.txt:
    * LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-computed.html:
    * LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-percentage-expected.html: Added.
    * LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-percentage-ref.html: Added.
    * LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-percentage.html: Added.
    * LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-valid-expected.txt:
    * LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-valid.html:
    * LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/text-underline-offset-expected.txt:
    * Source/WebCore/animation/CSSPropertyAnimation.cpp:
    (WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
    * Source/WebCore/css/CSSProperties.json:
    * Source/WebCore/css/ComputedStyleExtractor.cpp:
    (WebCore::textUnderlineOffsetToCSSValue):
    (WebCore::textDecorationThicknessToCSSValue):
    (WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
    * Source/WebCore/rendering/style/RenderStyle.h:
    * Source/WebCore/rendering/style/RenderStyleInlines.h:
    (WebCore::RenderStyle::initialTextUnderlineOffset):
    * Source/WebCore/rendering/style/TextUnderlineOffset.h:
    (WebCore::TextUnderlineOffset::createWithAuto):
    (WebCore::TextUnderlineOffset::createWithLength):
    (WebCore::TextUnderlineOffset::isAuto const):
    (WebCore::TextUnderlineOffset::isLength const):
    (WebCore::TextUnderlineOffset::length const):
    (WebCore::TextUnderlineOffset::resolve const):
    (WebCore::TextUnderlineOffset::TextUnderlineOffset):
    (WebCore::operator&lt;&lt;):
    (WebCore::TextUnderlineOffset::setLengthValue): Deleted.
    (WebCore::TextUnderlineOffset::lengthValue const): Deleted.
    (WebCore::TextUnderlineOffset::lengthOr const): Deleted.
    * Source/WebCore/style/InlineTextBoxStyle.cpp:
    (WebCore::computedUnderlineOffset):
    * Source/WebCore/style/StyleBuilderConverter.h:
    (WebCore::Style::BuilderConverter::convertTextUnderlineOffset):

Canonical link: <a href="https://commits.webkit.org/282611@main">https://commits.webkit.org/282611@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b58bd84d4809fef07541289a0bbb2cb8ce1aec02

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/63605 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42961 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/16202 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/67626 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/14213 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/65725 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50649 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/14493 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/51271 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9834 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66674 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/39868 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55081 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31900 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/36550 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/12457 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/13086 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/58056 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12799 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/69322 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/7552 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/12349 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/58546 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/7585 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/55170 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/58761 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14087 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6299 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/38782 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/39861 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/40973 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/39604 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -8349,7 +8349,7 @@
                 "converter": "TextUnderlineOffset",
-                "parser-grammar": "auto | <length>",
+                "parser-grammar": "auto | <length-percentage>",
                 "parser-grammar-comment": "The current specification has a <length-percentage>, not just <length>"
```
Should remove this comment.

zakariaridouh replied:
removed :)

mdubet commented on Source/WebCore/rendering/style/TextUnderlineOffset.h:
```diff
@@ -32,60 +32,84 @@ namespace WebCore {
+        }
     }
+private:
+    enum class Type : uint8_t {
```
Could be a `bool`

rreno commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -2811,6 +2804,49 @@ class PropertyWrapperBaselineShift final : public PropertyWrapper<SVGLengthValue
 };
 DEFINE_ALLOCATOR_WITH_HEAP_IDENTIFIER(PropertyWrapperBaselineShift);
 
+DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(PropertyWrapperTextUnderlineOffset);
```
I wish I hadn't made such fine-grained heap identifiers for every class in Animation. Might be worth a follow-up patch to make a generic "Animation" heap allocator then every Animation class can just use `WTF_MAKE_FAST_ALLOCATED_WITH_HEAP_IDENTIFIER(Animation)`.

If you don't want to do that, that's fine too I can do it some other time.

zakariaridouh replied:
Makes sense, yeah I'd love to pick it up. I can submit a patch for it after this one.

mdubet commented on Source/WebCore/rendering/style/RenderStyleInlines.h:
```diff
@@ -484,7 +484,7 @@ constexpr TextOverflow RenderStyle::initialTextOverflow() { return TextOverflow:
 constexpr TextSecurity RenderStyle::initialTextSecurity() { return TextSecurity::None; }
 inline StyleColor RenderStyle::initialTextStrokeColor() { return StyleColor::currentColor(); }
 constexpr OptionSet<TextTransform> RenderStyle::initialTextTransform() { return { }; }
-constexpr TextUnderlineOffset RenderStyle::initialTextUnderlineOffset() { return TextUnderlineOffset::createWithAuto(); }
```
It could stay `constexpr` it seems ?

zakariaridouh replied:
It can't, as `std::variant<>` doesn't support constexpr fully currently. Was recently added in C++23, so we can switch it back once we move to that.

mdubet commented on Source/WebCore/rendering/style/TextUnderlineOffset.h:
```diff
@@ -32,60 +32,84 @@ namespace WebCore {
+    static TextUnderlineOffset createWithAuto()
     {
-        return { std::nullopt };
+        return TextUnderlineOffset(Type::Auto);
```
Universal initialization would work here `{ Type::Auto }`

mdubet commented on Source/WebCore/rendering/style/TextUnderlineOffset.h:
```diff
@@ -32,60 +32,84 @@ namespace WebCore {
     }
 
-    constexpr float lengthOr(float defaultValue) const
+    constexpr bool operator==(const TextUnderlineOffset& other) const
```
We could keep the default generated comparison operator. But it makes me wonder if we want the underlying representation to be a `std::variant<Auto, Length>` actually ?

zakariaridouh replied:
Switched to using std::variant<Auto, Length>. Thanks.

mdubet commented on Source/WebCore/rendering/style/TextUnderlineOffset.h:
```diff
@@ -26,66 +26,85 @@
     }
 
-    constexpr float lengthOr(float defaultValue) const
+    bool operator==(const TextUnderlineOffset& other) const
```
We should just use the `= default` generated comparison

mdubet commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -2073,12 +2073,15 @@ static RefPtr<CSSValue> renderTextDecorationSkipToCSSValue(TextDecorationSkipInk
+    const Length& length = textUnderlineOffset.length();
+    if (length.isPercent())
+        return CSSPrimitiveValue::create(length.percent(), CSSUnitType::CSS_PERCENTAGE);
+    return ComputedStyleExtractor::zoomAdjustedPixelValueForLength(length, style);
```
We are changing the computed value of non percentage ones ? Why ?

weinig commented on Source/WebCore/rendering/style/TextUnderlineOffset.h:
```diff
@@ -26,66 +26,78 @@
-    constexpr float lengthOr(float defaultValue) const
+    bool operator==(const TextUnderlineOffset& other) const = default;
+private:
+    std::variant<Auto, Length> m_value;
```
Length has a type for Auto (`Length(LengthType::Auto)`). I don't think the variant is needed.

weinig commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -885,9 +885,15 @@ inline TextUnderlineOffset BuilderConverter::convertTextUnderlineOffset(BuilderS
     case CSSValueAuto:
         return TextUnderlineOffset::createWithAuto();
     default:
+        if (primitiveValue.isPercentage())
```
I think this entire converter can be simplified to:

```c++
return TextUnderlineOffset::createWithLength(BuilderConverter::convertLengthOrAuto(builderState, value));
```

zakariaridouh replied:
thank you!

nt1m commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -2078,7 +2078,10 @@ static Ref<CSSValue> textUnderlineOffsetToCSSValue(const TextUnderlineOffset& te
         return CSSPrimitiveValue::create(CSSValueAuto);
     ASSERT(textUnderlineOffset.isLength());
-    return CSSPrimitiveValue::create(textUnderlineOffset.lengthValue(), CSSUnitType::CSS_PX);
+    const Length& length = textUnderlineOffset.length();
```
```suggestion
    auto& length = textUnderlineOffset.length();
```

nt1m commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -2089,7 +2092,7 @@ static Ref<CSSValue> textDecorationThicknessToCSSValue(const RenderStyle& style,
 
     ASSERT(textDecorationThickness.isLength());
-    const auto& length = textDecorationThickness.length();
+    const Length& length = textDecorationThickness.length();
```
ditto

nt1m commented on Source/WebCore/rendering/style/TextUnderlineOffset.h:
I think we should file a follow up to remove the wrapper around `Length` entirely, since it's kind of pointless. `TextDecorationThickness` has it just because it supports the `from-font` value, but `TextUnderlineOffset` only supports `auto`

zakariaridouh replied:
Will open a followup

nt1m reviewed (approved):
Thanks! Please make sure to export the WPT :)

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-calc-expected.html:
```diff
@@ -0,0 +1,25 @@
+<title>text-underline-offset calc() support</title>
+<link rel="author" title="Zak Ridouh" href="mailto:zakr@apple.com" />
+<link rel="help" href="https://drafts.csswg.org/css-text-decor-4/#underline-offset" />
+<style type="text/css">
```
```suggestion
<style>
```

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-calc.html:
```diff
@@ -0,0 +1,46 @@
+<link rel="author" title="Zak Ridouh" href="mailto:zakr@apple.com" />
+<link rel="help" href="https://drafts.csswg.org/css-text-decor-4/#underline-offset" />
+<meta name="assert" content="Test checks whether text-underline-offset supports calc() values.">
+<style type="text/css">
```
```suggestion
<style>
```

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-calc-expected.html:
```diff
@@ -0,0 +1,25 @@
+<div>
+  <span class="underline ref">XXXXXX</span>
+</div>
+<br>
```
This `<br>` is not really useful

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-calc.html:
```diff
@@ -0,0 +1,46 @@
+  <span class="underline test4">X</span>
+  <span class="underline test5">X</span>
+</div>
+<br>
```
ditto

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-percentage.html:
```diff
@@ -0,0 +1,42 @@
+    <link rel="author" title="Apple" href="https://www.apple.com">
+    <link rel="help" href="https://drafts.csswg.org/css-text-decor-4/#underline-offset">
+    <style>
+        #main{
```
```suggestion
        #main {
```

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-percentage.html:
```diff
@@ -0,0 +1,42 @@
+            border-bottom: 2px solid purple;
+            display: flex;
+        }
+        #text, #norm{
```
```suggestion
        #text, #norm {
```

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-percentage.html:
```diff
@@ -0,0 +1,42 @@
+            position: relative;
+            margin-right: 10px;
+        }
+        #text{
```
```suggestion
        #text {
```

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-percentage.html:
```diff
@@ -0,0 +1,42 @@
+        #text{
+            text-underline-offset: 25%;
+        }
+        #norm{
```
```suggestion
        #norm {
```

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-percentage-expected.html:
```diff
@@ -0,0 +1,42 @@
+    <link rel="author" title="Apple" href="https://www.apple.com">
+    <link rel="help" href="https://drafts.csswg.org/css-text-decor-4/#underline-offset">
+    <style>
+        #main{
```
```suggestion
        #main {
```

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-percentage-expected.html:
```diff
@@ -0,0 +1,42 @@
+            border-bottom: 2px solid purple;
+            display: flex;
+        }
+        #text, #norm{
```
```suggestion
        #text, #norm {
```

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-percentage-expected.html:
```diff
@@ -0,0 +1,42 @@
+            position: relative;
+            margin-right: 10px;
+        }
+        #text{
```
```suggestion
        #text {
```

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-percentage-expected.html:
```diff
@@ -0,0 +1,42 @@
+        #text{
+            text-underline-offset: 5px;
+        }
+        #norm{
```
```suggestion
        #norm {
```

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-calc.html:
```diff
@@ -0,0 +1,46 @@
+<meta charset="UTF-8">
+<title>text-underline-offset calc() support</title>
+<link rel="author" title="Zak Ridouh" href="mailto:zakr@apple.com" />
+<link rel="help" href="https://drafts.csswg.org/css-text-decor-4/#underline-offset" />
```
This is missing a `<link rel="match" href="...">` pointing to a file that's identical to the `-expected.html` file, but ends with the `-ref.html` suffix.

Otherwise, this can't run on the WPT harness.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-percentage.html:
```diff
@@ -0,0 +1,42 @@
+    <title>Test case for text-underline-offset percentage support</title>
+    <link rel="author" title="Zak Ridouh" href="mailto:zakr@apple.com">
+    <link rel="author" title="Apple" href="https://www.apple.com">
+    <link rel="help" href="https://drafts.csswg.org/css-text-decor-4/#underline-offset">
```
ditto about `<link rel="match">`

zakariaridouh committed:
text-underline-offset should support percentages
https://bugs.webkit.org/show_bug.cgi?id=263431
rdar://117246233

Reviewed by Tim Nguyen.

This commit adds support for percentages in text-underline-offset [1].

Interpolation between different supported length types, such as em, %, and px
is supported and tested. Validate proper rendering with new tests. As specified
in the spec, % scales as a percentage of em/size of the font it inherits.

[1] https://drafts.csswg.org/css-text-decor-4/#valdef-text-underline-offset-percentage

    * LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/animations/text-underline-offset-interpolation-expected.txt:
    * LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/animations/text-underline-offset-interpolation.html:
    * LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-002-expected.html:
    * LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-002.html:
    * LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-calc-expected.html: Added.
    * LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-calc-ref.html: Added.
    * LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-calc.html: Added.
    * LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-computed-expected.txt:
    * LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-computed.html:
    * LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-percentage-expected.html: Added.
    * LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-percentage-ref.html: Added.
    * LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-percentage.html: Added.
    * LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-valid-expected.txt:
    * LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-underline-offset-valid.html:
    * LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/text-underline-offset-expected.txt:
    * Source/WebCore/animation/CSSPropertyAnimation.cpp:
    (WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
    * Source/WebCore/css/CSSProperties.json:
    * Source/WebCore/css/ComputedStyleExtractor.cpp:
    (WebCore::textUnderlineOffsetToCSSValue):
    (WebCore::textDecorationThicknessToCSSValue):
    (WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
    * Source/WebCore/rendering/style/RenderStyle.h:
    * Source/WebCore/rendering/style/RenderStyleInlines.h:
    (WebCore::RenderStyle::initialTextUnderlineOffset):
    * Source/WebCore/rendering/style/TextUnderlineOffset.h:
    (WebCore::TextUnderlineOffset::createWithAuto):
    (WebCore::TextUnderlineOffset::createWithLength):
    (WebCore::TextUnderlineOffset::isAuto const):
    (WebCore::TextUnderlineOffset::isLength const):
    (WebCore::TextUnderlineOffset::length const):
    (WebCore::TextUnderlineOffset::resolve const):
    (WebCore::TextUnderlineOffset::TextUnderlineOffset):
    (WebCore::operator<<):
    (WebCore::TextUnderlineOffset::setLengthValue): Deleted.
    (WebCore::TextUnderlineOffset::lengthValue const): Deleted.
    (WebCore::TextUnderlineOffset::lengthOr const): Deleted.
    * Source/WebCore/style/InlineTextBoxStyle.cpp:
    (WebCore::computedUnderlineOffset):
    * Source/WebCore/style/StyleBuilderConverter.h:
    (WebCore::Style::BuilderConverter::convertTextUnderlineOffset):

Canonical link: https://commits.webkit.org/282611@main
